Get var info from custom variables if not present in the tables

jvegasbsc opened the pull request:
Now the CMIP5 and 6 info classes will search for variable info also in our custom variables if they can not find a variable in their tables

This behaviour can be disabled by setting 'cmor_strict' to true at the config-developer.yml entry

For issue #725

mattiarighi commented:
Is this overlapping with #733 by @valeriupredoi ?

jvegasbsc commented:
Thanks Axel. I will take a look

jvegasbsc commented:
Ok. Just checked:

- If you provide the frequency in the recipe it works well. You can specify it at variable or dataset definition. I think this is something we can afford when using variables outside a data request. 

- After thatt, it will still fails because of `precipitation_flux_standard_error` being a non-CF standard name. Loaded constraint fails and everything fails after that. #638 should fix this because it no longer use the constraints as long as we remove the standard name from the definition

axel-lauer commented:
I merged this PR with #638 and with #733. When doing so, I run into (different) trouble, this time with _check_dim_names:

    if coordinate.generic_level:
    AttributeError: 'NoneType' object has no attribute 'generic_level'

This is somehow confusing as the variable prStderr does not have vertical levels.

Providing the key "frequency" in the recipe instead of the custom cmor table is a good solution and fine with me.

jvegasbsc commented:
Yes, this should be fixed in the last commit. A problem when reading the coordinates table for the custom cmor vars

axel-lauer commented:
My tests with the custom variable *prStderr* were now successful!

mattiarighi commented:
@axel-lauer are you using only this branch or also features from the other two branches addressing the same issues?

axel-lauer commented:
@mattiarighi For my latest test I used exclusively this branch, i.e. version2_default_var_info.

valeriupredoi commented:
@jvegasbsc just went through the code (no time to test it unfortunately, leaving on holidays tomorrow) but the code looks very good and it's a lot of great new additions! One quick question: will the `iris.load_raw()` call still be very restrictive on `constraint` or this PR solves that - I don't see any case treatment if the `standard_name` and/or `long_name` are missing - the case is needed data for files that don't have these names and just a `var_name` (which is in effect taken from the file name) is present

valeriupredoi commented:
see my changes in `_recipe.py` and `preprocessor/_io.py` in #733 - that's what I am on about :grin:

jvegasbsc commented:
>iris.load_raw() call still be very restrictive on constraint
#638 will remove completely the constraints. Nothing done on this to avoid mixing different changes.

The bad news is that @axel-lauer case needs both to be succesful

valeriupredoi commented:
cool cheers @jvegasbsc - I'll leave it to you then, happy holidays, guys :santa:

jvegasbsc commented:
In  fact, is not Iris list. It is CF conventions standard names list: http://cfconventions.org/standard-names.html. What we can do is to document this and encourage users to look for the standard name for their variable there and if it is not present, just leave it blank

axel-lauer commented:
The list of cf standard_names is not exclusive. I think things get confusing for users because obs4mips defines standard_names that did not make it (yet?) into this list of cf standard_names. If the need for leaving standard_name blank in such custom cmor tables will be well documented, fine with me even though this approach makes using the tool unnecessarily harder for users and is not user friendly at all.

axel-lauer commented:
@mattiarighi from my point of view, this PR would be ready for merging...

mattiarighi commented:
Let me quickly test it...

mattiarighi commented:
I get the following error when running `recipe_perfmetrics_CMIP5.yml`
```
esmvaltool._recipe.RecipeError: Missing keys {'frequency'} from variable lwcre in diagnostic lwcre
```
`lwcre` is a derived variable here.

jvegasbsc commented:
it is treated as a warning in CMIP5-like tables. as an error otherwise:

```python
        # Check other variable attributes that match entries in cube.attributes
        attrs = ('positive', )
        for attr in attrs:
            attr_value = getattr(self._cmor_var, attr)
            if attr_value:
                if attr not in self._cube.attributes:
                    # It is usually missing in CMIP5 data, so we only report
                    # a warning in that case
                    if self._cmor_var.table_type == 'CMIP5':
                        self.report_warning('{}: attribute {} not present',
                                            self._cube.var_name, attr)
                    else:
                        self.report_error('{}: attribute {} not present',
                                          self._cube.var_name, attr)
                elif self._cube.attributes[attr] != attr_value:
                    self.report_error(self._attr_msg, self._cube.var_name,
                                      attr, attr_value,
                                      self._cube.attributes[attr])
```

mattiarighi commented:
`rlut` is a standard CMIP5 variable, but probably the error is due to the fact that `rlut` is used to derive `lwcre`. Anyway, I think it should be always reported as a warning.

jvegasbsc commented:
Ok I will change it

mattiarighi commented:
This PR is fine with me, although I find the need to specify `frequency` for custom variables a bit confusing from the user's perspective.

mattiarighi commented:
@axel-lauer  before merging, we need to make sure that `frequency` is added to all existing recipes whenever a non-standard variable is used.

axel-lauer commented:
I added the 'frequency' key to the available recipes for derived variables. This shows how inconvenient this new requirement is as it requires duplicating information which is technically redundant as it was not needed before. Just as a side note, I am not very pleased with all these ideas about sticking to very strict standards when it comes to reading in data. In my opinion, the tool should remain as user friendly as possible even if that occasionally means that the implementation of some concepts will be less elegant.

jvegasbsc commented:
The frequency is required by one of the cmor checks, is just for that. If we are happy with some variables don't passing that check, I can change it so a warning is issued if frequency is not definde 

There is something that bothers me about all the definitions: we have frequency, realm and dimensions info and we still require the field info. This quite redundant and error prone. Maybe we should derive it and add it automatically if it is still used by some diagnostics (I am not using it for nothing in PRIMAVERA ones, and I think this is mostly useful for NCL ones in which you have less metadata)

valeriupredoi commented:
@jvegasbsc @mattiarighi - cheers for implementing this and cheers for changing the recipes; I took the recipes that had my name beside them and ran them through a set of tests just to be 100% all works fine; I found a rather strange error, here is the report:

```
PROBLEM REPORT 
---------------------------

Recipe affected:
recipe_autoassess_radiation_rms_cfMon_all.yml

version2_development:
------------------------------
fails: no

at cube load:
/home/users/valeriu/esmvaltool_var_test/rad_rms/recipe_autoassess_radiation_rms_cfMon_all_20190102_161458/preproc/radiation_cfMon_all_ISCCP_pp_rad_derive_var_clisccp/obs4mips_ISCCP_L3_V1.0_T4M_clisccp_1985-1988.nc
CONSTRAINTS: isccp_cloud_area_fraction

version2_default_var_info:
----------------------------------
fails: yes

clipped traceback:
  File "/home/users/valeriu/anaconda3Dec2018/envs/esmvaltool/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/preprocessor/_io.py", line 59, in load_cubes
    raise Exception('Can not load cubes from {0}'.format(files))
Exception: Can not load cubes from ('/group_workspaces/jasmin2/cmip6_prep_OLD/esmvaltool_users/valeriu/OBS_ISCCP/Tier1/ISCCP/clisccp_ISCCP_L3_V1.0_198707-198707.nc', ...

why: constraint is wrong

at cube load:
/home/users/valeriu/esmvaltool_var_test/rad_rms/recipe_autoassess_radiation_rms_cfMon_all_20190102_162113/preproc/radiation_cfMon_all_ISCCP_pp_rad_derive_var_clisccp/obs4mips_ISCCP_L3_V1.0_T4M_clisccp_1985-1988.nc
CONSTRAINTS: ISCCP_Low_Level_Medium_Thickness_Cloud_Area_Fraction
```
there is something wonky with the way the constraint is passed through only for the obs4mips case; the CMIP files are correctly constrained; also, all the other autoassess recipes work fine. If this issue is sorted out, I'd be happy with this merge :beer:

bouweandela commented:
> Comment by @mattiarighi: This PR is fine with me, although I find the need to specify `frequency` for custom variables a bit confusing from the user's perspective.

The frequency can simply be determined from the mip in almost all cases, e.g. by adding a function
```python
def mip_to_frequency(mip):
    """Try to guess the frequency from mip."""
    frequencies = (
        'subhr',
        'hr',
        '1hr',
        '3hr',
        '6hr',
        'day',
        'mon',
        'yr',
        'dec',
        'fx',
    )

    for frequency in frequencies:
        if frequency in mip or frequency.title() in mip:
            return frequency
```
and using that in _recipy.py to update the variable info dict if no frequency key is given in the recipe. That way users will rarely have to specify the frequency in the recipe.

> Comment by @jvegasbsc: There is something that bothers me about all the definitions: we have frequency, realm and dimensions info and we still require the field info. This quite redundant and error prone. 

I agree that it would be good to remove the `field` attribute from the recipes. I think we should stick with the standard esgf facets as much as possible in the recipe, because those are familiar to users, and indeed the information it provides is redundant. However, this is probably not something that should be addressed in this pull request.

bouweandela commented on esmvaltool/cmor/table.py:
```diff
@@ -25,18 +25,28 @@ def read_cmor_tables(cfg_developer):
-
         cmor_type = project.get('cmor_type', 'CMIP5')
-
+        cmor_strict = project.get('cmor_strict', False)
```
`True` would probably be a better default value, as it indicates that this is preferred.

valeriupredoi commented:
> @jvegasbsc @mattiarighi - cheers for implementing this and cheers for changing the recipes; I took the recipes that had my name beside them and ran them through a set of tests just to be 100% all works fine; I found a rather strange error, here is the report:
> 
> ```
> PROBLEM REPORT 
> ---------------------------
> 
> Recipe affected:
> recipe_autoassess_radiation_rms_cfMon_all.yml
> 
> version2_development:
> ------------------------------
> fails: no
> 
> at cube load:
> /home/users/valeriu/esmvaltool_var_test/rad_rms/recipe_autoassess_radiation_rms_cfMon_all_20190102_161458/preproc/radiation_cfMon_all_ISCCP_pp_rad_derive_var_clisccp/obs4mips_ISCCP_L3_V1.0_T4M_clisccp_1985-1988.nc
> CONSTRAINTS: isccp_cloud_area_fraction
> 
> version2_default_var_info:
> ----------------------------------
> fails: yes
> 
> clipped traceback:
>   File "/home/users/valeriu/anaconda3Dec2018/envs/esmvaltool/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/preprocessor/_io.py", line 59, in load_cubes
>     raise Exception('Can not load cubes from {0}'.format(files))
> Exception: Can not load cubes from ('/group_workspaces/jasmin2/cmip6_prep_OLD/esmvaltool_users/valeriu/OBS_ISCCP/Tier1/ISCCP/clisccp_ISCCP_L3_V1.0_198707-198707.nc', ...
> 
> why: constraint is wrong
> 
> at cube load:
> /home/users/valeriu/esmvaltool_var_test/rad_rms/recipe_autoassess_radiation_rms_cfMon_all_20190102_162113/preproc/radiation_cfMon_all_ISCCP_pp_rad_derive_var_clisccp/obs4mips_ISCCP_L3_V1.0_T4M_clisccp_1985-1988.nc
> CONSTRAINTS: ISCCP_Low_Level_Medium_Thickness_Cloud_Area_Fraction
> ```
> 
> there is something wonky with the way the constraint is passed through only for the obs4mips case; the CMIP files are correctly constrained; also, all the other autoassess recipes work fine. If this issue is sorted out, I'd be happy with this merge beer

I have just noticed that the cmor tables for `obs4mips` don't include cfMon - could this be the problem here? Do we have to add these ourselves?

bouweandela commented on esmvaltool/cmor/table.py:
```diff
@@ -25,18 +25,28 @@ def read_cmor_tables(cfg_developer):
     for table in cfg_developer:
         project = cfg_developer[table]
 
         table_path = project.get('cmor_tables', '')
```
It would probably be good to have esmvaltool/cmor/tables/<project_name> as the default path instead.

jvegasbsc replied:
I changed the code to make it a bit clearer

bouweandela commented on esmvaltool/cmor/tables/obs4mips/.github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,20 @@
+If you are *registering content* (RC) for obs4MIPs, please fill out the requested information below.   If you want to create an issue about something else, please delete the text below and title your issue as appropriate.  
```
I think it would be a cleaner to just copy the Tables directory into the esmvaltool repository, because that is all we need.

jvegasbsc replied:
I just cloned the repository to the path, as it is done with CMIP5 and CMIP6. I agree that this one is a bit more clutered, so I will leave only the json files

bouweandela replied:
Thanks. Did you do this already? Because there is still quite a lot of clutter around.

valeriupredoi commented:
I have noticed another (possible) problem in `recipe_autoassess_radiation_rms_Amon_obs.yml`:

```
problem: failed to load cube
file: /group_workspaces/jasmin4/esmeval/obsdata/Tier2/CERES-SYN1deg/OBS_CERES-SYN1deg_comp_Ed3A_T2Ms_rlutcs_200003-201609.nc
loading constraint (standard_name): toa_outgoing_longwave_flux_assuming_clear_sky
```

the constraint is indeed correct as per being asking for a valid CMOR name (as is in the CMIP5_Amon table for `rlutcs`) and the problem is the actual OBS file that does not have a standard_name but the question here is I thought this whole PR is meant to ease the case where these situations happen and introduce flexibility at cube load entry point; here, if the constraint would be relaxed to say, the `var_name`, the cube would be loaded OK

valeriupredoi commented:
@mattiarighi hope you don't mind I've updated your comment https://github.com/ESMValGroup/ESMValTool/pull/755#issuecomment-448939810 with results of my tests

bouweandela commented:
> I thought this whole PR is meant to ease the case where these situations happen and introduce flexibility at cube load entry point; here, if the constraint would be relaxed to say, the var_name, the cube would be loaded OK

No, that's in https://github.com/ESMValGroup/ESMValTool/pull/638. I think it would be best to finalize that PR before this one.

valeriupredoi commented:
@bouweandela thanks, man, I got mixed up - hey, it's still the first week of January :grin: OK, good then, but I still am quite confused by the other problem I am reporting, possibly because of the nonexistence of the `cfMon` table for `obs4mips`

jvegasbsc commented:
I think this is ready

valeriupredoi commented:
awesome @jvegasbsc gimme 2 seconds to test `recipe_autoassess_radiation_rms_cfMon_all.yml`

valeriupredoi commented:
am confused, now I can not load any of the CMOR tables for any of the derived variables...what am I missing here?

jvegasbsc commented:
Maybe I cleaned a bit to much. Let me take a look

valeriupredoi commented:
wha--what? :grinning: I was running in ...can not load CMIP5 CMOR table for `rtnt` which is not loading `toz` - I reran the `radiation_rms_Amon_all` recipe after I ran into issues with the `cfMon` one

jvegasbsc commented:
The issue should appear for every custom var you have. I assumed that you were testing the preprocessor derive namelist

valeriupredoi commented:
trouble in the jungle! what to do? Can I help?

jvegasbsc commented:
> trouble in the jungle! what to do? Can I help?
Test this update

valeriupredoi commented:
After putting `frequency: mon` in the example derive recipe, `KeyError: Dataset key modeling_realm must be specified` - same stuff for my autoassess recipes as well, man :cry:

valeriupredoi commented:
added `modeling_realm` in the dataset, and the `recipe_autoassess_radiation_rms_Amon_all.yml` starts chugging on but stops at:
`  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/preprocessor/_derive/__init__.py", line 87, in derive
    setattr(cube, attribute, variable[attribute])
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/iris/_cube_coord_common.py", line 127, in standard_name
    raise ValueError('%r is not a valid standard_name' % name)
ValueError: 'toa_net_downward_total_radiation' is not a valid standard_name`

jvegasbsc commented:
This is because we have defined invalid standard names in the custom definitions. This is something we have to avoid

valeriupredoi commented:
the `cfMon` recipe now stops right away (also after adding modeling_realm in dataset - do we really need that key?):
```
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_recipe.py", line 272, in _add_cmor_info
    table_entry.frequency = mip_info.frequency
AttributeError: 'NoneType' object has no attribute 'frequency'
```

jvegasbsc commented:
Try this update, @valeriupredoi

valeriupredoi commented:
good news - the Amon autoassess recipe wroked OK, the cfMon one is still giving us headaches:
```
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_recipe.py", line 774, in _get_preprocessor_task
    _add_cmor_info(variable, override=True, derive=True)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_recipe.py", line 278, in _add_cmor_info
    format(cmor_table, short_name, mip))
esmvaltool._recipe.RecipeError: Unable to load CMOR table 'obs4mips' for variable 'clisccp' with mip 'cfMon'
```
but that looks like the __good__ error since there is no `cfMon` table in `esmvaltool/cmor/tables/obs4mips/Tables`

valeriupredoi commented:
quick question: does `modeling_realm` now have to live in every dataset dictionary? Seems a bit of a shelf queen since we can probably do without it methinks

jvegasbsc commented:
> quick question: does modeling_realm now have to live in every dataset dictionary? Seems a bit of a shelf queen since we can probably do without it methinks

No, this should be fixed now. The modelling realm is available both in the variable and at the mip info, so if you get the correct variable info it should be added automatically

valeriupredoi commented:
the code fails for me if I don't add the `modeling_realm` in the recipe, in this branch, man

jvegasbsc commented:
> the code fails for me if I don't add the modeling_realm in the recipe, in this branch, man

Which recipe?

valeriupredoi commented:
sorry, didn't mention - all, including the `recipe_preprocessor_derive_test.yml` (for this one `frequency needs to be added as well, but after that it fails with needing `modeling_realm`) - you sure there's no typo in your code ie you actually spelled it correctly as `modelling_realm`?

valeriupredoi commented:
@jvegasbsc here's a patch, man: in `_data_finder.py` l.236 add the modeling realm info:
```
    # patch realm
    table = CMOR_TABLES[variable['cmor_table']].get_table(variable['mip'])
    realm = getattr(table.get(variable['short_name']), 'modeling_realm', None)
    variable['modeling_realm'] = realm if realm else table.realm
```
it works with this in :grin:

jvegasbsc commented:
Yes, I had trouble finding your error because I was using other convention for the path

valeriupredoi commented:
aha makes sense, shall we patch `_data-finder` like above?

jvegasbsc commented:
Try now. I am updating the variable info a step before looking for the files now

valeriupredoi commented:
my autoassess Amon recipe goes through but the preprocessor_derive_test is crapping up:
```
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_recipe.py", line 711, in <listcomp>
    for filename in _get_input_files(variable, config_user)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_recipe.py", line 578, in _get_input_files
    check_data_availability(input_files, variable)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_recipe.py", line 206, in check_data_availability
    "No input files found for variable {}".format(variable))
esmvaltool._recipe.RecipeError: No input files found for variable {'project': 'CMIP5', 'mip': 'Amon', 'exp': 'historical', 'start_year': 2000, 'end_year': 2005, 'field': 'T3M', 'derive': True, 'frequency': 'mon', 'force_derivation': False, 'short_name': 'tro3', 'diagnostic': 'diag1', 'preprocessor': 'default', 'dataset': 'GFDL-CM3', 'ensemble': 'r1i1p1', 'cmor_table': 'CMIP5', 'institute': ['NOAA-GFDL'], 'filename': '/home/users/valeriu/esmvaltool_var_test/cfmon_all/recipe_preprocessor_derive_test_20190116_150518/preproc/diag1_default_tro3/CMIP5_GFDL-CM3_Amon_historical_r1i1p1_T3M_tro3_2000-2005.nc', 'standard_name': 'mole_fraction_of_ozone_in_air', 'long_name': 'Mole Fraction of O3', 'units': '1e-9', 'modeling_realm': 'atmos atmosChem'}
```

valeriupredoi commented:
the modeling realm looks to be malformed ie 'atmos atmosChem'

valeriupredoi commented on esmvaltool/cmor/table.py:
```diff
@@ -508,7 +509,7 @@ def _read_variable(self, short_name, frequency):
             key, value = self._last_line_read
             if key in ('variable_entry', 'axis_entry'):
                 break
-            if key in ('dimensions', 'modeling_realm'):
```
keep the 'modeling_realm' and replace with `split()`

valeriupredoi commented on esmvaltool/cmor/table.py:
```diff
@@ -318,6 +318,7 @@ def read_json(self, json_data):
         self.valid_min = self._read_json_variable('valid_min')
         self.valid_max = self._read_json_variable('valid_max')
         self.positive = self._read_json_variable('positive')
+        self.modeling_realm = self._read_json_variable('modeling_realm')
```
replace with `self.modeling_realm = self._read_json_variable('modeling_realm').split()`

valeriupredoi commented on esmvaltool/cmor/table.py:
```diff
@@ -271,7 +271,7 @@ def __init__(self, table_type, short_name):
         """
         super(VariableInfo, self).__init__()
         self.table_type = table_type
-        self.modeling_realm = []
```
revert the change to `self.modeling_realm = []`

valeriupredoi commented:
made the thing work, man - see my comments

jvegasbsc commented:
Thanks!

valeriupredoi commented:
that stoopid `modeling_realm` must be an appendable list otherwise such fixes as https://github.com/ESMValGroup/ESMValTool/commit/a8d25ae732de57b561057dc705c58a96130ddcae would not work

valeriupredoi commented:
it runs as smooth as a greased-up pig :grin:  awesome @jvegasbsc :beer: Now, what are we gonna do about that obs4mips cfMon table, where to get it from?

jvegasbsc commented:
Add `cmor_strict = False` in the OBS4MIPS section of your config developer and let me know what happens

valeriupredoi commented:
`cmor_type: 'CMIP5'` and `cmor_strict: 'False'` = magic! the thing runs fine til the very end :grinning: Did notice an issue with multithreading, but that's not due to any code in here, it's pool writing iris cubes to the same output file that doesn't work on Jasmin : my obs4mips data is structured in monthly files and writing the output from each of these by each thread into a single file is not allowed on Jasmin ie parallel file operations pffft

```
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_main.py", line 215, in run
    conf = main(args)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_main.py", line 143, in main
    process_recipe(recipe_file=recipe, config_user=cfg)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_main.py", line 193, in process_recipe
    recipe.run()
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_recipe.py", line 1034, in run
    self.tasks, max_parallel_tasks=self._cfg['max_parallel_tasks'])
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_task.py", line 517, in run_tasks
    _run_tasks_parallel(tasks, max_parallel_tasks)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_task.py", line 562, in _run_tasks_parallel
    task.output_files = result.get()
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/multiprocessing/pool.py", line 644, in get
    raise self._value
OSError: [Errno -101] NetCDF: HDF error: b'/home/users/valeriu/esmvaltool_var_test/cfmon_all/recipe_autoassess_radiation_rms_cfMon_all_20190116_162604/preproc/radiation_cfMon_all_ISCCP_pp_rad_derive_var_clisccp/CMIP5_MRI-CGCM3_cfMon_amip_r1i1p1_T4M_clisccp_1985-1988.nc'
```
I will create an issue for this, strange I have not seen it until now

Anyway, am fully happy with this PR now :beer:

jvegasbsc commented:
Type should be CMIP6, as those tables are using json. Can you test with that?

jvegasbsc commented:
> is this issue taken care of the other PR (the fix raw cubes one)?

Yes

> for that cfMon recipe: cmor_type: 'CMIP5' and cmor_strict: 'False' works only, replacing with CMIP6 don't work

Can you send me the output when using CMIP6 ?

valeriupredoi commented:
great!

about the CMIP6 table:
```
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_recipe.py", line 711, in _get_derive_input_variables
    _add_cmor_info(var, override=True, derive=True)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a1-py3.6.egg/esmvaltool/_recipe.py", line 129, in _add_cmor_info
    format(cmor_table, short_name, mip))
esmvaltool._recipe_checks.RecipeError: Unable to load CMOR table 'obs4mips' for variable 'clisccp' with mip 'cfMon'
```

jvegasbsc commented:
It should be ready. Please @valeriupredoi , check the config-developer for the correct config and let me know what happens. Be aware that I have to add a variable to the custom cmor table and that you may encounter something similar (I could not test the full recipe)

mattiarighi commented:
I will test this once #638 is merged.

valeriupredoi commented:
> It should be ready. Please @valeriupredoi , check the config-developer for the correct config and let me know what happens. Be aware that I have to add a variable to the custom cmor table and that you may encounter something similar (I could not test the full recipe)

@jvegasbsc you are legend, man - both `cfMon` and `Amon_all` recipes worked straight out of the box! We need to remove all the `frequency:` keys in the recipes now

axel-lauer commented:
I just discovered a bug when implementing the derived variable xco2. Here is what I tried to do:
* models: variable is derived from standard cmor variables
* observations (obs4mips): variable exists --> no calculation needed
* a fix is applied to the observations because the longitudes in the original obs4mips file range from -180 to 180 instead of 0 to 360

Bug: the fix is applied twice(!) to the obs4mips dataset rendering it useless. I could picture this special case has simply not been considered in the current workflow.

bouweandela commented on esmvaltool/_recipe.py:
```diff
@@ -751,10 +745,10 @@ def _get_preprocessor_task(variables, profiles, config_user, task_name):
-    for variable in variables:
-        _add_cmor_info(variable)
+    else:
+        # Add CMOR info
```
Maybe this can be done before the `if variable.get('derive')` block and then the line ` _add_cmor_info(variable, derive=True)` can be removed from the function `_get_derive_input_variables`, because now this function has the unexpected side effect of changing `variables`.

jvegasbsc replied:
Done

bouweandela commented:
Have you looked at this code already? Is it still needed?
https://github.com/ESMValGroup/ESMValTool/blob/32ab4151fd0dc00bd1f8cf4ffd477e17b9c30de5/esmvaltool/preprocessor/_derive/__init__.py#L86-L89

jvegasbsc commented:
>I just discovered a bug when implementing the derived variable xco2. Here is what I tried to do:
    * models: variable is derived from standard cmor variables
    * observations (obs4mips): variable exists --> no calculation needed
    * a fix is applied to the observations because the longitudes in the original obs4mips file range from -180 to 180 instead of 0 to 360
Bug: the fix is applied twice(!) to the obs4mips dataset rendering it useless. I could picture this special case has simply not been considered in the current workflow.

I have updated your branch and checked that the fixes are only applied once. Also changed the fix to use the `intersection` function from Iris that should make it more efficient.

sloosvel commented:
We checked all the recipes, some of them fail because of diagnostic issues, but all the issues related to custom variable names have been solved. From our side, it is ready to merge.

mattiarighi commented:
@axel-lauer can you test again your cases and let us know if you are happy with this?

bouweandela commented on esmvaltool/_recipe.py:
```diff
@@ -744,16 +739,18 @@ def _get_preprocessor_task(variables, profiles, config_user, task_name):
 
         for derive_variables in derive_input.values():
+            for derive_variable in derive_variables:
+                _add_cmor_info(derive_variable)
```
I think you need
```suggestion
                _add_cmor_info(derive_variable, override=True)
```
otherwise it will not update the entries that are already present with the correct value

bouweandela commented:
The `frequency` keys still need to be removed from the recipes

jvegasbsc commented:
All done now

bouweandela commented on esmvaltool/recipes/recipe_autoassess_radiation_rms_cfMon_all.yml:
```diff
@@ -61,7 +61,7 @@ diagnostics:
         force_derivation: false
     additional_datasets:
-      - {dataset: ISCCP,  project: obs4mips,  level: L3,  version: V1.0,  start_year: 1985,  end_year: 1988, tier: 1}
+      - {dataset: ISCCP,  project: obs4mips, mip: cfMon, frequency: mon, level: L3,  version: V1.0,  start_year: 1985,  end_year: 1988, tier: 1}
```
```suggestion
      - {dataset: ISCCP,  project: obs4mips, level: L3,  version: V1.0,  start_year: 1985,  end_year: 1988, tier: 1}
```
I think this change is not needed?
